drivers/adc/ad463x/ad463x.c

```diff
@@ -529,23 +660,108 @@ int32_t ad463x_init(struct ad463x_dev **device,
 	axi_clkgen_remove(dev->clkgen);
 error_gpio:
 	no_os_gpio_remove(dev->gpio_resetn);
+	no_os_gpio_remove(dev->gpio_pgia_a0);
+	no_os_gpio_remove(dev->gpio_pgia_a1);
 error_dev:
 	no_os_free(dev);
 
-	return -1;
+	return -EINVAL;
+}
+
+/**
+ * @brief Find closest value of an array of intergers.
+ * @param target - Target/reference value.
+ * @param array - Input Array.
+ * @param size - array size.
+ * @return Gain control index, in case of success, negative number otherwise
+ */
+static int32_t find_closest(int32_t target, const int32_t array[], size_t size)
+{
+	if (size == 0) {
+		// Handle the case where the array is empty
+		return -EINVAL; // or any other appropriate value
+	}
+
+	int32_t closest_index = 0;
+	int32_t closest_diff = abs((int32_t)(target - array[0]));
+
+	for (size_t i = 1; i < size; ++i) {
+		uint64_t current_diff = abs((int32_t)(target - array[i]));
+
+		if (current_diff < closest_diff) {
+			closest_index = (int32_t)i;
+			closest_diff = current_diff;
+		}
+	}
+
+	return closest_index;
+}
+
+/**
+ * @brief Calculate the PGIA gain.
+ * @param gain_int - Interger part of the gain.
+ * @param gain_fract - Fractional part of the gain.
+ * @param vref - Reference Voltage.
+ * @param precision - Precision  value shifter.
+ * @return Gain control index, in case of success, -1 otherwise
+ */
+int32_t ad463x_calc_pgia_gain(int32_t gain_int, int32_t gain_fract,
+			      int32_t vref,
+			      int32_t precision)
+{
+	int32_t gain_idx;
+	uint64_t gain_nano, tmp;
+
+	gain_nano = (uint64_t)gain_int * NANO + gain_fract;
+	if (gain_nano > ADAQ4224_GAIN_MAX_NANO)
+		return -EINVAL;
+	tmp = gain_nano << precision;
+	tmp = NO_OS_DIV_ROUND_CLOSEST_ULL(tmp, NANO);
+	gain_nano = NO_OS_DIV_ROUND_CLOSEST_ULL(vref * 2, tmp);
+	gain_idx = find_closest(gain_nano, ad463x_gains,
+				NO_OS_ARRAY_SIZE(ad463x_gains));
+
+	return gain_idx;
+}
+
+/**
+ * @brief Set the PGIA gain.
+ * @param dev - Pointer to the device handler.
+ * @param gain_idx - Gain control index, consult values in ad463x_pgia_gain.
+ * @return 0 in case of success, -1 otherwise
+ */
+int32_t ad463x_set_pgia_gain(struct ad463x_dev *dev,
+			     uint8_t gain_idx)
+{
+	int32_t ret;
+	/** Check if gain available in the ADC */
+	if (!dev->has_pgia)
+		return -EINVAL;
+	/** Check if gain gain_idx is in the valid range */
+	if (gain_idx > 3)
+		return -EINVAL;
+
+	/** update gain index value in the device handler */
+	dev->pgia_idx = gain_idx;
+	/** Set A0 and A1 pins acoording to gain index value */
+	ret = no_os_gpio_set_value(dev->gpio_pgia_a0, gain_idx & NO_OS_BIT(0));
+	if (ret != 0)
+		return ret;
+	ret = no_os_gpio_set_value(dev->gpio_pgia_a1, (gain_idx>>1) & NO_OS_BIT(0));
```

Comment:
replace the bit operations with `no_os_field_get()`

Reply:
noted.